### PR TITLE
CI-Crossdock removed unnecessary arch builds in docker image

### PR DESCRIPTION
**Pull request by Wise-Wizard**
**Which problem is this PR solving?**

This PR addresses the issue [#5350 ](https://github.com/jaegertracing/jaeger/issues/5350).

Description:
Initialized QEMU only for linux/arm64 architecture.

**How was this change tested?**

The changes were tested by running the following command:

```bash
make test
```

**Checklist**

- [x] I have read [CONTRIBUTING_GUIDELINES.md](https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING_GUIDELINES.md)
- [x] I have signed all commits
- [x] I have added unit tests for the new functionality
- [x] I have run lint and test steps successfully
  - `for jaeger: make lint test`
  - `for jaeger-ui: yarn lint` and `yarn test`

### CODE REVIEW

File: .github/workflows/ci-crossdock.yml [hunk 1]
```diff
@@ -43,9 +43,13 @@ jobs:
     - name: Install tools
       run: make install-ci
 
-    - uses: docker/setup-qemu-action@68827325e0b33c7199eb31dd4e31fbe9023e06e3 # v3.0.0
+    # Conditionally setup QEMU only for arm64 Linux architecture
+    - name: Setup QEMU for arm64 Linux
+      if: runner.os == 'Linux' && runner.arch == 'arm64'
```
Comment (yurishkuro) [inline .github/workflows/ci-crossdock.yml:48]: This condition is always true, and the build is still going to go through all arch

**Review by yurishkuro — CHANGES_REQUESTED** [review]

File: .github/workflows/ci-crossdock.yml (hunk 1 above)
Comment (Wise-Wizard) [reply .github/workflows/ci-crossdock.yml]: Noted.

**Comment by yurishkuro** [comment]
Please use descriptive title for pr

File: scripts/build-crossdock.sh [hunk 2]
```diff
@@ -11,14 +11,19 @@ make build-and-run-crossdock
 if [[ "$BRANCH" == "main" ]]; then
   echo 'upload images to dockerhub/quay.io'
   REPO=jaegertracing/test-driver
-  IFS=" " read -r -a IMAGE_TAGS <<< "$(bash scripts/compute-tags.sh ${REPO})"
-  IMAGE_TAGS+=("--tag" "docker.io/${REPO}:${COMMIT}" "--tag" "quay.io/${REPO}:${COMMIT}")
+
+  TARGET_ARCH="linux/arm64"
+
+  IMAGE_TAGS=("--tag" "docker.io/${REPO}:${COMMIT}-${TARGET_ARCH}" "--tag" "quay.io/${REPO}:${COMMIT}-${TARGET_ARCH}")
```
Comment (yurishkuro) [inline scripts/build-crossdock.sh:17]: we do not include arch in the image tags. You can see in the production images that the same tag could represent multiple architectures, e.g.

image

**Comment by yurishkuro** [comment]
https://github.com/jaegertracing/jaeger/actions/runs/8921061123/job/24500365309?pr=5356#step:10:197
```
docker buildx build -t localhost:5000/baseimg_alpine:latest --push \
	--build-arg root_image=alpine:3.16 \
	--build-arg cert_image=alpine:3.16 \
	--platform=linux/amd64,linux/s390x,linux/ppc64le,linux/arm64 \
	docker/base
```
still builds many platforms

File: scripts/build-crossdock.sh (hunk 2 above)
Comment (Wise-Wizard) [reply scripts/build-crossdock.sh]: Yes, thanks for clarifying. Tried a different approach now, went through the QEMU documentation, and noticed we can specify platforms as an input to QEMU, if we want only specific platforms to be built? Is this approach right?

**Comment by yurishkuro** [comment]
the approach was outlined in the ticket - to use the same mechanism that is used for not building all platforms for all other images when not on main branch https://github.com/jaegertracing/jaeger/blob/14aa43d4929ef344df722081aafd1a76c93c4f22/.github/workflows/ci-docker-build.yml#L53

**Comment by Wise-Wizard** [comment]
HI, found the issue. The actual docker image being built is from another Makefile inside the docker folder.
Screenshot 2024-05-03 021523
So, we have to pass the platform as an argument from the build-crossodck.sh file itself, to only build the amd-64 architecture instead?

**Comment by Wise-Wizard** [comment]
> Found the issue. The actual docker image being built is from another Makefile inside the docker folder. Screenshot 2024-05-03 021523 So, we have to pass the platform as an argument from the build-crossodck.sh file itself, to only build the amd-64 architecture instead?

On hardcoding, the platform variable as linux/amd-64, it is not building any other architectures!

File: .github/workflows/ci-crossdock.yml [hunk 3]
```diff
@@ -45,12 +45,18 @@ jobs:
 
     - uses: docker/setup-qemu-action@68827325e0b33c7199eb31dd4e31fbe9023e06e3 # v3.0.0
 
-    - name: Build, test, and publish crossdock image
+    - name: Build crossdock image (main branch only)
+      if: github.ref_name == 'main'
```
Comment (yurishkuro) [inline .github/workflows/ci-crossdock.yml:49]: we don't need these ifs, unlike main binaries, crossdock always can be built on one architecture only

File: docker/Makefile [hunk 4]
```diff
@@ -14,13 +14,13 @@ create-baseimg: prepare-docker-buildx
 	docker buildx build -t $(BASE_IMAGE) --push \
 		--build-arg root_image=$(ROOT_IMAGE) \
 		--build-arg cert_image=$(CERT_IMAGE) \
-		--platform=$(PLATFORMS) \
+		--platform=linux/amd64 \
```
Comment (yurishkuro) [inline docker/Makefile:17]: again, if you need to change the behavior, just override the PLATFORMS var, don't change the actual target code

File: .github/workflows/ci-crossdock.yml (hunk 3 above)
Comment (Wise-Wizard) [reply .github/workflows/ci-crossdock.yml:49]: Yes, I will be removing that.
The main change of this commit was only the Makefile

File: docker/Makefile (hunk 4 above)
Comment (Wise-Wizard) [reply docker/Makefile:17]: Yes, doing that right now.

**Comment by yurishkuro** [comment]
The issue has been solve in #5410. Closing this PR, thanks for trying.

**Comment by Wise-Wizard** [comment]
> The issue has been solve in #5410. Closing this PR, thanks for trying.

I had made the same change, and was going to push it just now. He built upon my work anyways.
No Worries :)

**Comment by varshith257** [comment]
> > The issue has been solve in #5410. Closing this PR, thanks for trying.
> 
> I had made the same change, and was going to push it just now. He built upon my work anyways.
> No Worries :)

@Wise-Wizard I haven't taken your work. You can see my closed PR of my attempts to rectify the root cause and @yurishkuro also knows it. And this suggestion made by yurishkuro before here suggested for you to override variable. You can follow up in my closed PR :)

**Comment by varshith257** [comment]
@yurishkuro PTAL . This is something going wrong in our organization. If someone working on issue please don't make others to work on them until closes/merge PR and simply waste of the efforts of others.  

I think it can better handled by assignment of issues to contributer to avoid this conflicts

**Comment by Wise-Wizard** [comment]
> > > The issue has been solve in #5410. Closing this PR, thanks for trying.
> > 
> > 
> > I had made the same change, and was going to push it just now. He built upon my work anyways.
> > No Worries :)
> 
> @Wise-Wizard I haven't taken your work. You can see my closed PR of my attempts to rectify the root cause and @yurishkuro also knows it. And this suggestion made by yurishkuro to me before here suggested for you to override variable. You can follow up in my closed PR :)

No worries Varshith, but as you mentioned you weren't able to find the root cause. I found the root cause and was on the verge of completing it, so you should have just let me go through with it. Again. it's all in a good cause and the issue was resolved, it's all that matters :)

**Pull request closed without merging**